hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java

```diff
@@ -775,8 +915,8 @@ public void writeJsonFormattedField(StringBuilder stringBuilder, AbstractGeneric
         };
         ALL[FieldKind.PORTABLE.getId()] = new FieldKindBasedOperations() {
             @Override
-            public Object readObject(GenericRecord genericRecord, String fieldName) {
-                return ((InternalGenericRecord) genericRecord).getObject(fieldName);
+            public Object readAsLeafObjectOnQuery(InternalGenericRecord genericRecord, String fieldName) {
+                return genericRecord.getObject(fieldName);
```

Comment:
So, we are trying to read an object out of InternalGenericRecord, right? That feels a bit weird, why do we need it?

Reply:
This is used from Query when the user's query ends with an object. 
`SELECT * FROM map WHERE x.y.z.employee > ? `

In this case, we have to construct `Employee` class to be able to compare on the server. This method is called in this case.